### ISSUE TITLE
Memory leak in model.fit

### ISSUE THREAD

**Issue by nkoerb** [tensorflow/tensorflow#37505]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information** 
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): 
minimal working example

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): 
windows-server 2016
- TensorFlow installed from (source or binary): 
conda
- TensorFlow version (use command below): 
tf 2.1.0
- Python version: 
3.7
- CUDA/cuDNN version: 
CUDA 10.1
- GPU model and memory:
K80 - 24Gb

**Describe the current behavior**
memory use increases with consecutive training runs, probably related to #35524, #33030, #35124, #35835
side note: I do not understand the warning but this seems to be handled in #37500
**Describe the expected behavior**
memory should stay constant

**Standalone code to reproduce the issue** 
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```
from tensorflow.keras.datasets import cifar10
import tensorflow.keras.callbacks as callbacks
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.callbacks import LearningRateScheduler
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.layers import Input, Conv2D, GlobalAveragePooling2D, Activation, Dense
from tensorflow.keras.models import Model
import tensorflow.keras.utils as kutils
import numpy as np
import psutil
import gc 


batch_size = 128
epochs = 5
num_classes = 10


def buildmodel():
    img_input = Input(shape=(32, 32, 3))

    x = Conv2D(16, (3, 3), padding='same')(img_input)
    x = Activation("relu")(x)
    x = Conv2D(16, (3, 3), padding='same')(x)
    x = Activation("relu")(x)
    x = GlobalAveragePooling2D()(x)
    prediction = Dense(num_classes,activation='softmax', name = 'classifier') (x)  
    model = Model(inputs=img_input, outputs=prediction)
    return model

(trainX, trainY), (testX, testY) = cifar10.load_data()
mean = np.mean(trainX, axis=0)
std = np.std(trainX)

trainX = trainX.astype('float32')
trainX = (trainX - mean) / std
testX = testX.astype('float32')
testX = (testX - mean) / std

trainY = kutils.to_categorical(trainY)
testY = kutils.to_categorical(testY)

generator = ImageDataGenerator()
generator.fit(trainX)
val_generator = ImageDataGenerator()

for i in range(10):
    tf.keras.backend.clear_session()
    model = buildmodel()
    sgd = SGD(lr=0.1, momentum=0.9, nesterov=True)
    model.compile(loss="categorical_crossentropy", optimizer=sgd, metrics=["acc"])
    model.fit(generator.flow(trainX, trainY, batch_size=batch_size), epochs=epochs, validation_data=val_generator.flow(testX, testY, batch_size = batch_size),verbose=0, workers = 20)
    print('memory usesd: ' + str(psutil.virtual_memory().used // 1e6))
    gc.collect()
```

output:
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
memory usesd: 38012.0
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
memory usesd: 38563.0
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
memory usesd: 39288.0
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
memory usesd: 40005.0
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
memory usesd: 40730.0
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
memory usesd: 41490.0
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
memory usesd: 42216.0
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
memory usesd: 42937.0
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
memory usesd: 43659.0
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
WARNING:tensorflow:sample_weight modes were coerced from
  ...
    to  
  ['...']
memory usesd: 44403.0

**Comment by Saduf2019** [comment]
I have replicated the code shared above, please find the [gist here](https://colab.sandbox.google.com/gist/Saduf2019/1df4ffa08b3cd2491fa3395fe4815dac/37505.ipynb)

**Comment by howl-anderson** [comment]
I am trying to fix this issue, I do find the issue on Colab, but on my MacBook Pro (13-inch, 2017, Two Thunderbolt 3 ports), there seems no such memory leak issue.

Here is the output of the program (warning msgs are removed);
```text
memory usesd: 8968.0
memory usesd: 9009.0
memory usesd: 8402.0
memory usesd: 8299.0
memory usesd: 8257.0
memory usesd: 7761.0
memory usesd: 8642.0
memory usesd: 8833.0
memory usesd: 9008.0
memory usesd: 9267.0
```

**Comment by nkoerb** [comment]
@howl-anderson : do you have something running in parallel? Your memory usage seems not to be constant over time

**Comment by howl-anderson** [comment]
@nkoerb Yes, it's a little weird, I will test again without open other applications on Mac and test it on Ubuntu 18.04 too.

**Comment by howl-anderson** [comment]
@nkoerb New test on Mac (without open other applications):

```text
memory usesd: 6879.0
memory usesd: 7259.0
memory usesd: 7620.0
memory usesd: 7966.0
memory usesd: 8323.0
memory usesd: 8694.0
memory usesd: 9029.0
memory usesd: 9392.0
memory usesd: 9486.0
memory usesd: 9480.0
```

**Comment by nkoerb** [comment]
@howl-anderson looks leaky to me

**Comment by howl-anderson** [comment]
@nkoerb Agree! I am will try to fix this later.

**Comment by howl-anderson** [comment]
This issue likely relates to #37653

**Comment by DanyelJei** [comment]
Hey, I was having the same issue. 
My current workaround is to save the model before deleting it and clearing the backend after each training iteration. Then I reload the model before calling fit again. 
I'm no longer experiencing the memory leak with this workaround.

**Comment by nkoerb** [comment]
@DanyelJei how do you clear the backend, because I also used clear_session() and garbage collector? And what exactly do you mean with 'training iteration': full training, epoch or single iteration on a batch?

**Comment by chrisffm** [comment]
Have a similar issue with increasing memory usage when using the ImageDataGenerator and the fit function

**Comment by dxxd116** [comment]
I have the same memory leaking problem with model.fit when using dataset from generator. 
Hopefully it will be fixed some time soon.

**Comment by alexpyattaev** [comment]
Confirm a memory leak, similar symptoms. Seems to eat about 4 Mbytes/second on my workload. tf2.2.0-1, arch 64bit.

**Comment by dansitu** [comment]
Also experiencing this issue

**Comment by grudloff** [comment]
Hey, same issue here on Windows machine. Seems to be related to multiple model fits over generators, it clearly seems like a memory leak from tensorflow side, deleting all variables and calling garbage collector won't fix the issue. Managed to evade the problem by setting `workers = 1`.

**Comment by luvwinnie** [comment]
> Hey, same issue here on Windows machine. Seems to be related to multiple model fits over generators, it clearly seems like a memory leak from tensorflow side, deleting all variables and calling garbage collector won't fix the issue. ~~Managed to evade the problem by setting `workers = 1`.~~
> 
> 
> 
> Edit: I thought setting `workers=1` would fix the issue as described in a stackoverflow post I can't find right now. Trying now with `workers=0`.
> 
> Edit2: Nope, seems this doesn't work either.
> 
> Edit3: Managed to make it work with `workers=0` by letting each iteration of training in the loop to be in a function, so that the variables are local, followed by a call to `tf.keras.backend.clear_sesion()` and then `gc.collect()`.

Hi, could you provide some information or example code of fix with your edit3?

**Comment by ysyyork** [comment]
i noticed the validation_data is actually causing the memory leak. not sure if it's because of my implementation cus I'm using dataset as a wrapper of my multiprocessing data loader. I'm using `tf.data.Dataset.from_generator(generator_fun, ...)` What happens to me is that if you dig into the source code, you will actually see that the validation data and the train x in the fit function are treated differently. 

In my case, my `generator_fun` is actually getting called every epoch for validation set but for training x, it's not called every time. Then everytime my `generator_fun` generated a new generator which causes the memory leak. I verified this behavior by digging down to the `model.fit` source code. I'm on tf 2.2.0 and you can see  below under [tf.python.keras.engine.training.py](https://github.com/tensorflow/tensorflow/blob/64dbe97a4d5a9a2cc69cacae7762157cd201a1d2/tensorflow/python/keras/engine/training.py#L859) :

```
        # Run validation.
        if validation_data and self._should_eval(epoch, validation_freq):
          val_x, val_y, val_sample_weight = (
              data_adapter.unpack_x_y_sample_weight(validation_data))
          val_logs = self.evaluate(
              x=val_x,
              y=val_y,
              sample_weight=val_sample_weight,
              batch_size=validation_batch_size or batch_size,
              steps=validation_steps,
              callbacks=callbacks,
              max_queue_size=max_queue_size,
              workers=workers,
              use_multiprocessing=use_multiprocessing,
              return_dict=True)
          val_logs = {'val_' + name: val for name, val in val_logs.items()}
          epoch_logs.update(val_logs)
```
So the dataset passed in will be put into the evaluate function where it probably called the as_numpy_iterator() and that caused my `generator_fun` got called again and again. However the training generator didn't show this behavior. Then I modified my generator_fun so that every time it's called it's not creating a new instance but reusing the same generator that has already been created, then the memory leak is gone. 


 I'm not sure how the ImageDataGenerator.flow_from_directory work but maybe you can check that. Also i notice they modified this part of code on master. I haven't tried nightly yet. maybe they already fixed it. Just a different perspective.

**Comment by DanlanChen** [comment]
Is there any solution? I am still facing it now..

**Comment by siarencjusz** [comment]
I ran into a similar problem. Workaround which seems to work for me is to force garbage collector:


```
import gc
...
model.fit(...)
gc.collect()
```

**Comment by jvishnuvardhan** [comment]
@nkoerb Related to performance, there were some significant improvements in the recent version. I ran your code with `tf-nightly` and I don't see any improvement in the memory as much as you have noticed. When I ran your code with `tf-nightly`, i see the following ouput.

Downloading data from https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz
170500096/170498071 [==============================] - 2s 0us/step
memory usesd: 3981.0
memory usesd: 3981.0
memory usesd: 3976.0
memory usesd: 3975.0
memory usesd: 3975.0
memory usesd: 3973.0
memory usesd: 3973.0
memory usesd: 3970.0

Can you please verify once with `tf-nightly` and close the issue if this was resolved for you. Thanks!

**Comment by nkoerb** [comment]
@jvishnuvardhan Thanks for your help and great work, memory is constant with tf-nightly! 
As a side note: When I remove either `gc.collect()` or `tf.keras.backend.clear_session()` memory is leaking, so both commands are needed for constant memory usage when performing consecutive trainings.

**Comment by krenerd** [comment]
Is the solve of this issue applied to the current tensorflow release 2.3.2?

**Comment by jvishnuvardhan** [comment]
@krenerd Can you please open a new issue with a simple standalone code to reproduce your issue? Thanks! You can refer this issue in the new issue. Thanks

**Comment by Sherry40931** [comment]
I encountered similar problems with keras.fit + tf.dataset, for me the problem is `prefetch`. I changed
```python
datasets = datasets.prefetch(dim * batch_size)
```
to
```python
datasets = datasets.prefetch(tf.data.experimental.AUTOTUNE)
```
and the memory leak problem is solved, hope it helps.

**Comment by SysuJayce** [comment]
> I encountered similar problems with keras.fit + tf.dataset, for me the problem is `prefetch`. I changed
> 
> ```python
> datasets = datasets.prefetch(dim * batch_size)
> ```
> 
> to
> 
> ```python
> datasets = datasets.prefetch(tf.data.experimental.AUTOTUNE)
> ```
> 
> and the memory leak problem is solved, hope it helps.

@Sherry40931 hello, Im facing the same problem, and Im using tf.dataset.from_tensor_slices(), when training with keras.model.fit() distributedly with TF 2.4.1, memory usage keeps increasing. Below is the main procedure of my dataset part.
```
ds = tf.data.Dataset.from_tensor_slices()
ds = ds.map().cache()
ds = ds.shuffle(reshuffle_each_iteration=True).batch()
keras.model.fit(ds)
```
Can u share some advice for me ?

**Comment by krenerd** [comment]
@SysuJayce The following code solved the problem for me. 

```
import gc

...

tf.keras.backend.clear_session()
gc.collect()
```

**Comment by SysuJayce** [comment]
> @SysuJayce The following code solved the problem for me.
> 
> ```
> import gc
> 
> ...
> 
> tf.keras.backend.clear_session()
> gc.collect()
> ```
@krenerd Hi, thx for the reply. Where should I add this part in my code?

**Comment by krenerd** [comment]
@SysuJayce In my code, the memory leakage happened after every iteration of reading images and training one epoch epoch on it. So I appended the code after every iteration. Although this is not a fundamental fix to this problem we experienced, it fixed my issue. 
```
for epoch in range(epochs):
  data,labels=dataset.load_train(...)
  self.model.fit(data,labels)
  tf.keras.backend.clear_session()
  gc.collect()
```

**Comment by SysuJayce** [comment]
> @SysuJayce In my code, the memory leakage happened after every iteration of reading images and training one epoch epoch on it. So I appended the code after every iteration. Although this is not a fundamental fix to this problem we experienced, it fixed my issue.
> 
> ```
> for epoch in range(epochs):
>   data,labels=dataset.load_train(...)
>   self.model.fit(data,labels)
>   tf.keras.backend.clear_session()
>   gc.collect()
> ```

@krenerd unfortunately, I tried your code but memory leak still exists

**Comment by Sherry40931** [comment]
hi @SysuJayce, I think the order of your functions is correct, could you provide the parameters/functions you send to map, cache and batch?

**Comment by SysuJayce** [comment]
> hi @SysuJayce, I think the order of your functions is correct, could you provide the parameters/functions you send to map, cache and batch?

@Sherry40931  hi, actually, I did a dataset benchmark test, and found no memory leak. So maybe the problem is within the fit() / predict()?

I use TextVectorization layer to convert text to vector in my map(), and cache() has no parameter, and batch(batch_size=32).
```
def vectorize_text(text, label):
    text = tf.expand_dims(text, -1)
    return vectorize_layer(text), label
```

Below is the dataset memory test snippet
```
for it in ds.repeat():
    used_mem = psutil.virtual_memory().used
    print(f"used memory: {used_mem / 1024 / 1024} Mb")
```

**Comment by atotev** [comment]
https://fantashit.com/linearly-increasing-memory-with-use-multiprocessing-and-keras-sequence/#comment-254237

**Comment by Ribe12** [comment]
> https://fantashit.com/linearly-increasing-memory-with-use-multiprocessing-and-keras-sequence/#comment-254237

This is the only thing that worked for me. I was experincing a memory leak each epoch during the validation phase with a data generator (it didn't happened if I didn't validate). Apparently the fit() function doesn't handle well the data generator for validation when using multiprocessing.

Spent weeks trying to solve it, tried with gc.collect(), K.clear_session(), among other; but no luck. By wrapping the generator with GeneratorEnqueuer (https://www.tensorflow.org/versions/r2.3/api_docs/python/tf/keras/utils/GeneratorEnqueuer) to handle multiprocessing before passing it to fit() (workers=1 and multiprocessing=False must be set in fit()), the memory leak was gone.

I am using tf 2.3.2

**Comment by jayagami** [comment]
> > https://fantashit.com/linearly-increasing-memory-with-use-multiprocessing-and-keras-sequence/#comment-254237
> 
> This is the only thing that worked for me. I was experiencing a memory leak each epoch during the validation phase with a data generator (it didn't happened if I didn't validate). Apparently the fit() function doesn't handle well the data generator for validation when using multiprocessing.
> 
> Spent weeks trying to solve it, tried with gc.collect(), K.clear_session(), among other; but no luck. By wrapping the generator with GeneratorEnqueuer (https://www.tensorflow.org/versions/r2.3/api_docs/python/tf/keras/utils/GeneratorEnqueuer) to handle multiprocessing before passing it to fit() (workers=1 and multiprocessing=False must be set in fit()), the memory leak was gone.
> 
> I am using tf 2.3.2

Worked for me too. 
Can't believe that this problem still remains unsolved in 2021 on tf2.5

**Comment by AndreyStille** [comment]
I found the solution!
Not use this piece of shit library written by fucking unprincipled dumbasses and morons. What they can do great is just do fucking plastic smiles, pass difficult useless interviews and lick ass to their bosses from the mose fascist organization of all times. Please die from cancer!
Pytorch forever

**Comment by krenerd** [comment]
This bug seems to be fixed in `tf.nightly` as of June 28th. 
https://github.com/tensorflow/tensorflow/issues/50379

**Comment by jiunyen-ching** [comment]
> Then I modified my generator_fun so that every time it's called it's not creating a new instance but reusing the same generator that has already been created, then the memory leak is gone.

@ysyyork Thank you for this insight! May I know what modifications have you made to your generator function to reuse the same generator?

**Comment by ysyyork** [comment]
> > Then I modified my generator_fun so that every time it's called it's not creating a new instance but reusing the same generator that has already been created, then the memory leak is gone.
> 
> @ysyyork Thank you for this insight! May I know what modifications have you made to your generator function to reuse the same generator?

@jiunyen-ching I used a singleton pattern to implement this generator.

**Comment by shir994** [comment]
Can anyone advise how I check since which version this bug is implemented in a stable release? Can't find any info on the releases page:
https://github.com/tensorflow/tensorflow/releases?page=1

**Comment by nkoerb** [comment]
I think it was solved from tf version 2.3 -> 2.4

**Comment by 27rg5** [comment]
Once this problem is solved, I can try to make this code multi-gpu friendly if we need additional speedup.
